### PR TITLE
vmconstantpool updates for supporting OpenJDK MethodHandles

### PR DESCRIPTION
**Pull request by babsingh**
In order to support **OpenJDK MethodHandles**, the below mentioned class,
field and static method references are introduced in **vmconstantpool.xml**.

**Class references:**
1. java/lang/invoke/MemberName (j.l.i.MemberName)
2. j.l.i.MethodType
3. j.l.i.MethodTypeHelper
4. j.l.i.MethodTypeForm
5. j.l.i.LambdaForm
6. j.l.i.MethodHandle
7. j.l.i.MethodHandleHelper
8. j.l.i.WrongMethodTypeException

**Field references:**
1. From j.l.i.MemberName,
   - name="clazz" signature="Ljava/lang/Class;"
   - name="name" signature="Ljava/lang/String;"
   - name="type" signature="Ljava/lang/Object;"
   - name="flags" signature="I"
   - name="resolution" signature="Ljava/lang/Object;"
2. From j.l.i.MethodType,
   - name="ptypes" signature="[Ljava/lang/Class;"
   - name="rtype" signature="Ljava/lang/Class;"
   - name="form" signature="Ljava/lang/invoke/MethodTypeForm;"
   - name="methodDescriptor" signature="Ljava/lang/String;"
3. From j.l.i.MethodTypeForm,
   - [JDK14+]  name="parameterSlotCount" signature="S"
   - [JDK8-11] name="argCounts" signature="J"
   - "parameterSlotCount" is an alias for "argCounts"
4. From j.l.i.LambdaForm,
   - name="vmentry" signature="Ljava/lang/invoke/MemberName;"
5. From j.l.i.MethodHandle,
   - name="form" signature="Ljava/lang/invoke/LambdaForm;"
   - name="type" signature="Ljava/lang/invoke/MethodType;"

**Static method references:**
1. From j.l.i.MethodTypeHelper,
   - name="vmResolveFromMethodDescriptorString"
2. From j.l.i.MethodHandleHelper.
   - name="sendResolveMethodHandle"
   - name="resolveMethodHandle"
   - name="resolveInvokeDynamic"
   - name="resolveConstantDynamic"
   - name="constructorPlaceHolder"

These new additions are only enabled if the opt_openjdkMethodhandle flag
is specified. Currently, this flag is disabled.

Co-authored-by: Jack Lu <Jack.S.Lu@ibm.com>
Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
@fengxue-IS can you please review this pull request?

**Comment by fengxue-IS** [comment]
CI build failed, looking at output, I don't think we can define a class/field/method twice with the same name but under different flags.

Need to either modify the sourcetools to support having multiple flags for one entry or putting common ones under the `default` flag.

PS: you will have to add
```
#ifdef J9VM_OPT_OPENJDK_METHODHANDLE
	flags |= JCL_RTFLAG_OPT_OPENJDK_METHODHANDLE;
#endif
```
to **runtime/jcl/common/stdinit.c::computeJCLRuntimeFlags()**

**Comment by babsingh** [comment]
@fengxue-IS Thanks for refreshing my memory about the unsupported `vmconstantpool` features. For the time being, I have taken the NO FLAG approach for the two duplicate field references, which resolves the errors. I have also added `JCL_RTFLAG_OPT_OPENJDK_METHODHANDLE`. Let me consult others about the unsupported `vmconstantpool` features.

File: runtime/oti/vmconstantpool.xml [hunk 1]
```diff
@@ -340,6 +348,30 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
...
+	<fieldref class="java/lang/invoke/LambdaForm" name="vmentry" signature="Ljava/lang/invoke/MemberName;" flags="opt_openjdkMethodhandle"/>
+	<fieldref class="java/lang/invoke/MethodHandle" name="form" signature="Ljava/lang/invoke/LambdaForm;" flags="opt_openjdkMethodhandle"/>
+		
+	<!--
+	Field references used by both OpenJ9 and OpenJDK MethodHandles are mentioned below.
+	Multiple flags or multiple definitions under different flags are unsupported for field references.
+	So, the below field references are defined without a flag field.
+	These field references are only needed for opt_methodHandle and opt_openjdkMethodhandle flags.
+	-->
+	<fieldref class="java/lang/invoke/MethodType" name="methodDescriptor" signature="Ljava/lang/String;"/>
```
Comment (DanHeidinga) [inline runtime/oti/vmconstantpool.xml:372]: Another alternative is to add an `opt_methodHandleCommon` flag that is required by both `opt_methodHandle` and `opt_openjdkMethodhandle`.

This comment can be moved to the description of that flag and it makes it clear (and enforced) the dependency between them

Comment (babsingh) [reply runtime/oti/vmconstantpool.xml:372]: This should work. I will add another flag for common dependencies between OpenJ9 and OpenJDK MethodHandles.

Comment (babsingh) [reply runtime/oti/vmconstantpool.xml]: Added the `opt_methodHandleCommon` flag.

**Comment by babsingh** [comment]
**Update:** I added the `opt_methodHandleCommon` flag, and disabled all OpenJ9 dependencies which are not needed by OpenJDK MethodHandles. Then, I tested these changes locally with `opt_openjdkMethodhandle` enabled. `MethodHandle.resolveConstantDynamic` is disabled for OpenJDK MethodHandles. But, it goes through a `javaLookupMethod` operation during init, which fails. This shouldn't be possible when `opt_methodHandle` is disabled. When I delete the method ref for  `MH.resolveConstantDynamic`, everything works magically. This behaviour is only seen for `MethodHandle.resolveConstantDynamic`. No other `opt_methodHandle` ref shows this behaviour when `opt_openjdkMethodhandle` is enabled and `opt_methodHandle` is disabled. @fengxue-IS @DanHeidinga Would you know why this is happening?

```
<staticmethodref class="java/lang/invoke/MethodHandle" name="resolveConstantDynamic" signature="(JLjava/lang/String;Ljava/lang/String;J)Ljava/lang/Object;" versions="11-" flag="opt_methodHandle"/>
```

**Comment by DanHeidinga** [comment]
I looked in the code for the jclcinit.c and it shows we're only checking the flags on the ROMClass to decide whether to attempt the method or field resolve:
https://github.com/eclipse/openj9/blob/c75ec5e05edb3c7d78de4d1568cf58a1f6cdaee2/runtime/jcl/common/jclcinit.c#L314-L319

This is likely a limitation in the vm constant pool tool - to really make this work it would need to ensure there is a J9ROMClassRef with the set of `runtimeFlags` that each field / method needs.  This would result in the same class being included multiple times in the vmconstantpool - once for each combo of flags required by the classref, the fieldrefs, and the methodrefs.  Which means we might have multiple ways to get to the classref at runtime depending on which macro was used. 

I think this is a limitation of the tool - the flags on the methods and fields must match those on the class - or the resolution will happen anyway.

It's probably not worth adding this support to the tool so I'd suggest going back to your previous approach.

**Comment by babsingh** [comment]
Merging `MethodHandle.resolveConstantDynamic` [currently OpenJ9 only] into `MethodHandleHelper.resolveConstantDynamic` [OpenJDK -> common] with `ifdefs` and deleting `MethodHandle.resolveConstantDynamic` should resolve this issue. This can be done for all the MH static method refs.

**Comment by babsingh** [comment]
These changes are ready for a review. I have decided to keep the `opt_methodHandleCommon` flag. 

**Follow up changes:**
1. When the `MethodTypeHelper` PR is ready, move the `MethodTypeHelper` staticmethodref under the `opt_methodHandleCommon` flag, and delete the corresponding `MethodType` staticmethodref.
2. When the `MethodHandleHelper` PR is ready, move the `MethodHandleHelper` staticmethodrefs under the `opt_methodHandleCommon` flag, and delete the corresponding `MethodHandle` staticmethodrefs.

**Comment by babsingh** [comment]
@DanHeidinga can you please review this PR?

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk8

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux jdk11

**Comment by DanHeidinga** [comment]
Jenkins test sanity win jdk15

**Pull request merged**